### PR TITLE
[query] Add series metadata label limit header for trimming response

### PR DESCRIPTION
**Pull request by robskillington**
**What this PR does / why we need it**:

Adds ability to trim series metadata / label metadata via header.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
NONE
```

### CODE REVIEW

File: src/query/api/v1/handler/prom/read.go [hunk 1]
```diff
@@ -148,20 +148,25 @@ func (h *readHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
-				zap.Error(err), zap.String("query", query),
-				zap.Bool("instant", h.opts.instant))
-		}
+	limited := &handleroptions.ReturnedDataLimited{
+		Limited:     returnedDataLimited.Limited,
+		Series:      returnedDataLimited.Series,
+		TotalSeries: returnedDataLimited.TotalSeries,
+		Datapoints:  returnedDataLimited.Datapoints,
+	}
+	err = handleroptions.AddResponseHeaders(w, resultMetadata, fetchOptions, limited, nil)
```
Comment (wesleyk) [inline src/query/api/v1/handler/prom/read.go:161]: so we're always adding the response header now, 👍 

cc @rallen090

File: src/query/api/v1/handler/prometheus/remote/match.go [hunk 2]
```diff
@@ -103,9 +106,33 @@ func (h *PromSeriesMatchHandler) ServeHTTP(w http.ResponseWriter, r *http.Reques
...
 	}
 
-	handleroptions.AddResponseHeaders(w, meta, opts)
+	var (
+		noopWriter = ioutil.Discard
+		renderOpts = prometheus.RenderSeriesMetadataOptions{
+			ReturnedSeriesMetadataLimit: opts.ReturnedSeriesMetadataLimit,
+		}
+	)
+	renderResult, err := prometheus.RenderSeriesMatchResultsJSON(noopWriter, results, renderOpts)
```
Comment (rallen090) [inline src/query/api/v1/handler/prometheus/remote/match.go:115]: Can we leave comments on these noop render calls to explain the purpose is to apply the limiting first so we can write the header before the actual body writing?

File: src/query/api/v1/handler/prometheus/common.go [hunk 3]
```diff
@@ -347,6 +375,12 @@ func renderDefaultTagCompletionResultsJSON(
 		jw.BeginObjectField("values")
 		jw.BeginArray()
 		for _, value := range tag.Values {
```
Comment (rallen090) [inline src/query/api/v1/handler/prometheus/common.go:377]: So I'm assuming it is intentional here that we still write the key/value field/array even once over limit, and that the arrays are just empty?

File: src/query/api/v1/handler/prometheus/common.go [hunk 4]
```diff
@@ -339,6 +361,12 @@ func renderDefaultTagCompletionResultsJSON(
 	jw.BeginArray()
 
 	for _, tag := range results {
+		total += len(tag.Values)
+		if opts.ReturnedSeriesMetadataLimit > 0 && rendered >= opts.ReturnedSeriesMetadataLimit {
+			limited = true
+			continue
```
Comment (rallen090) [inline src/query/api/v1/handler/prometheus/common.go:367]: Ok I see - so we `continue` here so we can still count `total`

**Review by rallen090 — APPROVED** [review]
LGTM

File: src/query/api/v1/handler/prometheus/remote/match.go (hunk 2 above)
Comment (robskillington) [reply src/query/api/v1/handler/prometheus/remote/match.go:115]: Yeah sg.

**Pull request merged**